Improve the Mockup documentation (#1345)
plone/mockup: README.md

There are several options to integrate Mockup.

1.  It comes pre-installed with Plone 6 Classic UI.

> 2.  You can install and compile a bundle. See [insert anchor to specific section, as "below" is vague].

Reviewer: I've linked that to the `Development` section.

    On this port our demo and documentation pages are served.

## Mockup overview

Mockup is a JavaScrip UI library which provides widgets, apps and functionality for the Plone Classic UI variant.
It is based on Patternslib(XXX) and provides it's functionality through so-called Patterns which are initialized and get active if a triggering CSS selector is found in the DOM tree.
For example, the related items widget is initialized on a form input field with the CSS class `pat-relateditems`.
The configuration is done via data attributes, in case of the related items pattern it's an attribute called `data-pat-relateditems`.

> The data strucutre of the configuration can be a JSON string or CSS declaration like key-value pairs separated by a semicolon.

Reviewer: Hm ... this needs maybe more clarification. The key/values are separated with colons but the k/v pairs are separated with semicolon:

```
key: value;
key: value;
...
```

> The data strucutre of the configuration can be a JSON string or CSS declaration like key-value pairs separated by a semicolon.

Reviewer 2: I updated my suggestion accordingly. Please let me know. Thank you!

## Mockup overview

Mockup is a JavaScript UI library which provides widgets, apps, and functionality for the Plone Classic UI frontend.

> It is based on Patternslib(XXX), and provides its functionality through patterns.

Reviewer: This should show the current used Patternslib version from `package.json`. Is it possible to extract that in Markdown automatically?

> It is based on Patternslib(XXX), and provides its functionality through patterns.

Reviewer 2: We currently manually update the version of packages in our `conf.py`.

https://github.com/plone/documentation/blob/6.0/docs/conf.py#L306-L327

The dict could be extended automatically. Then in the MyST file:

https://github.com/plone/documentation/blob/e011a6df4e8444a2f7ae59ec9678aa7ed60bf374/docs/install/manage-add-ons-packages.md?plain=1#L64

> It is based on Patternslib(XXX), and provides its functionality through patterns.

Reviewer 2: Done in [5f5f92f](https://github.com/plone/mockup/pull/1345/commits/5f5f92f5dd9168540356ee0eb92874400d23c0b5)

The ``@@plonejsi18n`` view is called one time for a given domain and language and the result is cached in localStorage for 24 hours.
The only way to test the new translations is to restart the instance to update the mo file from the po file, and then to purge the localStorage to trigger a new download of the translations.

### Styleguide

-   Tab: 4 spaces for everything except for HTML and XML files (2 Spaces) and the Makefile (TAB).
    This rule is defined in the [.editorconfig file](./.editorconfig).

### Commit style guide

We automatically generate the changelog from the commit messages following the [Conventional Commits specification](https://www.conventionalcommits.org/).
Changelog generation is done by the [conventional changelog plugin](https://github.com/release-it/conventional-changelog/) for [release-it](https://github.com/release-it/release-it).
This is enforced via a pre-commit hook managed by [husky](https://github.com/typicode/husky).

**And this is how you use it:**

We have 4 different types of changelog entries:

-   Breaking Changes (type: `breaking`),
-   Features (type: `feat`),
-   Bug Fixes (type: `fix`),

> -   Maintenance (type: `maint`).

Reviewer 2: Other repos also have:

```
-   Documentation (type: `documentation`)
```

`.doc` opens in Word and other office apps by default, and this file extension is explicit.

Is that something you want to add?

Reviewer: We could add it in mokcup but we currently reuse `release-it` config from [@patternslib/dev](https://github.com/Patternslib/dev/blob/main/.release-it.js) so it should be extended there first before we mention it here.

@thet what do you think?

Reviewer 3: maybe open an issue for this feature and merge this PR

Reviewer 2: Issue created: https://github.com/plone/mockup/issues/1373

This translation helper that calls the ``@@plonejsi18n`` view defined in plone.app.content to generate a JSON of the translations from the mo file.
The ``@@plonejsi18n`` view is called one time for a given domain and language and the result is cached in localStorage for 24 hours.
The only way to test the new translations is to restart the instance to update the mo file from the po file, and then to purge the localStorage to trigger a new download of the translations.

> ### Styleguide

Reviewer 2: ```suggestion
## Style guide
```

-   Tab: 4 spaces for everything except for HTML and XML files (2 Spaces) and the Makefile (TAB).
    This rule is defined in the [.editorconfig file](./.editorconfig).

### Commit style guide

We automatically generate the changelog from the commit messages following the [Conventional Commits specification](https://www.conventionalcommits.org/).
Changelog generation is done by the [conventional changelog plugin](https://github.com/release-it/conventional-changelog/) for [release-it](https://github.com/release-it/release-it).
This is enforced via a pre-commit hook managed by [husky](https://github.com/typicode/husky).

**And this is how you use it:**

We have 4 different types of changelog entries:

-   Breaking Changes (type: `breaking`),
-   Features (type: `feat`),
-   Bug Fixes (type: `fix`),
-   Maintenance (type: `maint`).

> We can group commits in the changelog via a scope - or grouping.

Reviewer 2: ```suggestion
We can group commits in the changelog via a scope or grouping.
```

> Let's follow a convention and use these groupings - but the grouping is optional and any other group name can be used.

Reviewer 2: ```suggestion
Let's follow a convention and use these groupings, but the grouping is optional and any other group name can be used.
```

-   "Dependencies" for upgrading package.json dependencies.
-   "Docs" for documentation.
-   "Build" for everything related to the release workflow, Webpack and building bundles.
-   "Cleanup" for cleaning up or reformatting code.

> -   "pat PATTERNNAME" for changes to individual patterns in src/pat.

Reviewer 2: ```suggestion
-   "pat PATTERNNAME" for changes to individual patterns in `src/pat`.
```

> -   "core MODULENAME" for changes to core modules in src/core.

Reviewer 2: ```suggestion
-   "core MODULENAME" for changes to core modules in `src/core`.
```

> A commit message with a changelog-grouping must be structured like so: `TYPE(GROUP): MESSAGE`.

Reviewer 2: ```suggestion
A commit message with a changelog grouping must be structured like so: `TYPE(GROUP): MESSAGE`.
```

Without grouping: `TYPE: MESSAGE`

> If the commit message doesn't follow this convention it won't be included in the changelog.

Reviewer 2: ```suggestion
If the commit message doesn't follow this convention, then it won't be included in the changelog.
```

To bypass the pre-commit hook, use the git `-n` switch.

> Example: `git commit yarn.lock -m"yarn install." -n`.

Reviewer 2: ```suggestion
Example: `git commit yarn.lock -m "yarn install." -n`.
```

If you are working on a component like the structure pattern (pat-structure), use `pat structure` as a group.

Examples:

Add a feature to the structure pattern:

> ```

Reviewer 2: ````suggestion
```shell
````

> git commit src/pat/structure -m"feat(pat structure): Add feature to cook some coffee"

Reviewer 2: ```suggestion
git commit src/pat/structure -m "feat(pat structure): Add feature to cook some coffee"
```

```

Cleanup task:

> ```

Reviewer 2: ````suggestion
```shell
````

> git commit -am"maint(Cleanup): Remove whitespace from documentation."

Reviewer 2: ```suggestion
git commit -am "maint(Cleanup): Remove whitespace from documentation."
```

```

or without a grouping:

> ```

Reviewer 2: ````suggestion
```shell
````

> git commit -am"maint: Remove unnecessary file from root directory."

Reviewer 2: ```suggestion
git commit -am "maint: Remove unnecessary file from root directory."
```

```

---

**Note**

Please keep commits on the `yarn.lock` file or any other auto-generated code seperate.

> Just commit them seperately with `git commit yarn.lock -m"yarn install" -n`.

Reviewer 2: ```suggestion
Just commit them seperately with `git commit yarn.lock -m "yarn install" -n`.
```

> Having them seperately reduces the effort when merging or rebasing branches where a merge conflict can easily happen.

Reviewer 2: ```suggestion
Keeping them seperate reduces the effort when merging or rebasing branches where a merge conflict can easily happen.
```

> In such cases you can just withdraw your changes on the yarn.lock file or remove those commits and re-install with `yarn install` at the end of a successful merge or rebase.

Reviewer 2: ```suggestion
In such cases you can just withdraw your changes on the `yarn.lock` file, or remove those commits and reinstall with `yarn install` at the end of a successful merge or rebase.
```

Let's follow a convention and use these groupings, but the grouping is optional and any other group name can be used.

-   "Dependencies" for upgrading package.json dependencies.
-   "Docs" for documentation.
-   "Build" for everything related to the release workflow, Webpack and building bundles.
-   "Cleanup" for cleaning up or reformatting code.
-   "pat PATTERNNAME" for changes to individual patterns in `src/pat`.
-   "core MODULENAME" for changes to core modules in `src/core`.

A commit message with a changelog grouping must be structured like so: `TYPE(GROUP): MESSAGE`.
Without grouping: `TYPE: MESSAGE`

If the commit message doesn't follow this convention, then it won't be included in the changelog.
To bypass the pre-commit hook, use the git `-n` switch.
Example: `git commit yarn.lock -m "yarn install." -n`.

If you are working on a component like the structure pattern (pat-structure), use `pat structure` as a group.

Examples:

Add a feature to the structure pattern:

```shell
git commit src/pat/structure -m "feat(pat structure): Add feature to cook some coffee"
```

Cleanup task:

```shell
git commit -am "maint(Cleanup): Remove whitespace from documentation."

> 

Reviewer 2: ```suggestion
```
